### ISSUE TITLE
Probing Liveness/Readyness is failing when tls-config is enabled.

### ISSUE THREAD

**Issue by ybettan** [observatorium/observatorium#36]
Apparently, when enabling tls-config, `/-/healty` and `/-/ready` endpoints require some client-certificate as well as `write` and `read` endpoints. Therefore, when deployed on k8s/OCP the pod fails to deploy.

`oc describe` output:
```
Events:
  Type     Reason     Age               From                         Message
  ----     ------     ----              ----                         -------
  Normal   Scheduled  <unknown>         default-scheduler            Successfully assigned observatorium/observatorium-xyz-observatorium-api-gateway-78d458758c-pvtbx to crc-w6th5-master-0
  Normal   Pulled     1m                kubelet, crc-w6th5-master-0  Container image "quay.io/observatorium/observatorium:master-2020-04-16-v0.1.0-2-g2a4aa00" already present on machine
  Normal   Created    1m                kubelet, crc-w6th5-master-0  Created container observatorium-api-gateway
  Normal   Started    1m                kubelet, crc-w6th5-master-0  Started container observatorium-api-gateway
  Warning  Unhealthy  20s               kubelet, crc-w6th5-master-0  Liveness probe failed: HTTP probe failed with statuscode: 400
  Warning  Unhealthy  3s (x7 over 33s)  kubelet, crc-w6th5-master-0  Readiness probe failed: HTTP probe failed with statuscode: 400
```

`oc logs` output:
```
level=debug name=observatorium ts=2020-04-23T12:52:53.512151038Z caller=main.go:94 msg="maxprocs: Leaving GOMAXPROCS=[6]: CPU quota undefined"
level=info name=observatorium ts=2020-04-23T12:52:53.51238815Z caller=main.go:109 msg="starting observatorium"
level=info name=observatorium ts=2020-04-23T12:52:53.512610768Z caller=config.go:27 protocol=HTTP msg="enabling server side TLS"
level=info name=observatorium ts=2020-04-23T12:52:53.535028952Z caller=config.go:68 protocol=HTTP msg="server TLS client verification enabled"

**Comment by brancz** [comment]
liveness and readiness probes need to be re-configured to do TLS if you configure it that way. Could you share your whole setup though so we can understand the context of this better?

**Comment by ybettan** [comment]
I have talked about it with @kakkoyun and I think @metalmatze should be updated as well.
When enabling mTLS, observatorium doesn't differ between `/-/healthy`, `/-/ready` endpoints and `read`/`write` endpoints, therefore, probing those endpoints requires client-certificate.

When k8s/OCP tries to deploy api-gateway's pod, it probs `healthy` and `ready` endpoints without supplying any valid client-certificate therefore, in my opinion, the pod fails to deploy.

You can see all my setup here:
https://github.com/ybettan/configuration/tree/devel

I use `./deploy.sh` to deploy the app.

I have also tried to update `containers.livenessProbe.httpGet.scheme` and `containers.readinessProbe.httpGet.scheme ` to `HTTPS` but I get other TLS error:

`oc describe` output:
```
Events:
  Type     Reason     Age                From                         Message
  ----     ------     ----               ----                         -------
  Normal   Scheduled  <unknown>          default-scheduler            Successfully assigned observatorium/observatorium-xyz-observatorium-api-gateway-558db99b77-7476s to crc-w6th5-master-0
  Normal   Pulled     1m                 kubelet, crc-w6th5-master-0  Container image "quay.io/observatorium/observatorium:master-2020-04-16-v0.1.0-2-g2a4aa00" already present on machine
  Normal   Created    1m                 kubelet, crc-w6th5-master-0  Created container observatorium-api-gateway
  Normal   Started    1m                 kubelet, crc-w6th5-master-0  Started container observatorium-api-gateway
  Warning  Unhealthy  27s (x2 over 57s)  kubelet, crc-w6th5-master-0  Liveness probe failed: Get https://10.128.0.198:8080/-/healthy: remote error: tls: protocol version not supported
  Warning  Unhealthy  1s (x13 over 1m)   kubelet, crc-w6th5-master-0  Readiness probe failed: Get https://10.128.0.198:8080/-/ready: remote error: tls: protocol version not supported
```

`oc logs` output:
```
level=debug name=observatorium ts=2020-04-27T07:59:56.770803531Z caller=main.go:94 msg="maxprocs: Leaving GOMAXPROCS=[6]: CPU quota undefined"
level=info name=observatorium ts=2020-04-27T07:59:56.771315674Z caller=main.go:109 msg="starting observatorium"
level=info name=observatorium ts=2020-04-27T07:59:56.771391034Z caller=config.go:27 protocol=HTTP msg="enabling server side TLS"
level=info name=observatorium ts=2020-04-27T07:59:56.786128745Z caller=config.go:68 protocol=HTTP msg="server TLS client verification enabled"
level=info name=observatorium ts=2020-04-27T07:59:56.786712991Z caller=prober.go:85 msg="changing probe status" status=healthy
level=info name=observatorium ts=2020-04-27T07:59:56.786772385Z caller=server.go:133 msg="starting the HTTP server" address=0.0.0.0:8080
level=info name=observatorium ts=2020-04-27T07:59:56.786822232Z caller=prober.go:61 msg="changing probe status" status=ready
2020/04/27 08:00:01 http: TLS handshake error from 10.128.0.1:47814: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:06 http: TLS handshake error from 10.128.0.1:47938: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:11 http: TLS handshake error from 10.128.0.1:48104: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:16 http: TLS handshake error from 10.128.0.1:48234: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:21 http: TLS handshake error from 10.128.0.1:48402: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:25 http: TLS handshake error from 10.128.0.1:48506: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:26 http: TLS handshake error from 10.128.0.1:48530: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:31 http: TLS handshake error from 10.128.0.1:48692: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:36 http: TLS handshake error from 10.128.0.1:48812: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:41 http: TLS handshake error from 10.128.0.1:48974: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:46 http: TLS handshake error from 10.128.0.1:49100: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:51 http: TLS handshake error from 10.128.0.1:49264: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:55 http: TLS handshake error from 10.128.0.1:49376: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:00:56 http: TLS handshake error from 10.128.0.1:49400: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:01 http: TLS handshake error from 10.128.0.1:49566: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:06 http: TLS handshake error from 10.128.0.1:49688: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:11 http: TLS handshake error from 10.128.0.1:49840: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:16 http: TLS handshake error from 10.128.0.1:49970: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:21 http: TLS handshake error from 10.128.0.1:50124: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:25 http: TLS handshake error from 10.128.0.1:50234: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:26 http: TLS handshake error from 10.128.0.1:50262: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:31 http: TLS handshake error from 10.128.0.1:50432: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:36 http: TLS handshake error from 10.128.0.1:50562: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:41 http: TLS handshake error from 10.128.0.1:50718: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:46 http: TLS handshake error from 10.128.0.1:50836: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:51 http: TLS handshake error from 10.128.0.1:50996: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:55 http: TLS handshake error from 10.128.0.1:51100: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:01:56 http: TLS handshake error from 10.128.0.1:51126: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:01 http: TLS handshake error from 10.128.0.1:51288: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:06 http: TLS handshake error from 10.128.0.1:51420: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:11 http: TLS handshake error from 10.128.0.1:51584: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:16 http: TLS handshake error from 10.128.0.1:51708: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:21 http: TLS handshake error from 10.128.0.1:51880: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:25 http: TLS handshake error from 10.128.0.1:51984: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:26 http: TLS handshake error from 10.128.0.1:52008: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:31 http: TLS handshake error from 10.128.0.1:52178: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:36 http: TLS handshake error from 10.128.0.1:52306: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:41 http: TLS handshake error from 10.128.0.1:52474: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:46 http: TLS handshake error from 10.128.0.1:52608: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:51 http: TLS handshake error from 10.128.0.1:52776: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:55 http: TLS handshake error from 10.128.0.1:52876: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:02:56 http: TLS handshake error from 10.128.0.1:52896: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:01 http: TLS handshake error from 10.128.0.1:53062: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:06 http: TLS handshake error from 10.128.0.1:53192: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:11 http: TLS handshake error from 10.128.0.1:53356: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:16 http: TLS handshake error from 10.128.0.1:53478: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:21 http: TLS handshake error from 10.128.0.1:53644: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:25 http: TLS handshake error from 10.128.0.1:53762: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:26 http: TLS handshake error from 10.128.0.1:53782: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:31 http: TLS handshake error from 10.128.0.1:53954: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:36 http: TLS handshake error from 10.128.0.1:54084: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:41 http: TLS handshake error from 10.128.0.1:54240: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:46 http: TLS handshake error from 10.128.0.1:54370: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:51 http: TLS handshake error from 10.128.0.1:54526: tls: client offered only unsupported versions: [303 302 301]
2020/04/27 08:03:55 http: TLS handshake error from 10.128.0.1:54638: tls: client offered only unsupported versions: [303 302 301]
level=info name=observatorium ts=2020-04-27T08:03:55.022202467Z caller=main.go:127 msg="caught interrupt"
level=warn name=observatorium ts=2020-04-27T08:03:55.022320713Z caller=prober.go:70 msg="changing probe status" status=not-ready reason=null
level=info name=observatorium ts=2020-04-27T08:03:55.02234639Z caller=server.go:163 msg="shutting down internal server"
```

**Comment by brancz** [comment]
That looks like the client is trying to do TLS1.2 whereas the server is expecting min TLS1.3. How about instead of http probes we just change them to TCP probes?

**Comment by ybettan** [comment]
Using TCP worked. Now the pod is deployed.
But now I see that metrics cannot be delivered with TLS because of https://github.com/coreos/prometheus-operator/issues/3118.

**Comment by brancz** [comment]
We also need to think about gathering metrics from the gateway itself (which we may want to put on a different listener so we can handle authentication separately).

**Comment by squat** [comment]
yes, @metalmatze and I agreed to simplify this issue by moving the all debug and metrics handlers to a new http server on a different port

**Comment by metalmatze** [comment]
The internal server was merged with #39. Moving forward the metrics are on that server, which only supports HTTP.
